openff/bespokefit/executor/utilities/redis.py

```diff
@@ -5,12 +5,87 @@
 import shutil
 import subprocess
 import time
-from typing import IO, Optional, Union
+from typing import IO, Dict, Optional, Tuple, Union
 
 import redis
 
+from openff.bespokefit.executor.services import current_settings
 
-def is_redis_available(host: str, port: int = 6379) -> bool:
+__REDIS_VERSION: int = 1
+__CONNECTION_POOL: Dict[Tuple[str, int, Optional[int], bool], redis.Redis] = {}
+
+
+class RedisNotConfiguredError(BaseException):
+    """An exception raised when connecting to a redis server that doesn't appear to
+    have been configured by `openff-bespokefit`.
+    """
+
+
+class RedisBadConfigurationError(BaseException):
+    """An exception raised when connecting to a redis server that doesn't appear to
+    have been correctly configured for use with `openff-bespokefit`.
+    """
+
+
+def expected_redis_config_version() -> int:
+    return __REDIS_VERSION
+
+
+def connect_to_default_redis(validate: bool = True) -> redis.Redis:
+    """Connects to a redis server using the settings defined by the
+    `BEFLOW_REDIS_ADDRESS`, `BEFLOW_REDIS_PORT` and `BEFLOW_REDIS_PORT` settings.
+    """
+
+    settings = current_settings()
+
+    return connect_to_redis(
+        settings.BEFLOW_REDIS_ADDRESS,
+        settings.BEFLOW_REDIS_PORT,
+        settings.BEFLOW_REDIS_DB,
+        validate=validate,
+    )
+
+
+def connect_to_redis(
+    host: str, port: int, db: int, validate: bool = True
+) -> redis.Redis:
+    """Connects to a redis server using the specified settings."""
+
+    connection_key = (host, port, db, validate)
+
+    if connection_key in __CONNECTION_POOL:
```

Comment:
Is there any danger with multiple workers using the same connection object to redis?

Reply:
I'm not 100% certain, but I'm pretty sure it won't affect anything as the connection shouldn't really be doing anything other than pinging the redis server!